pallene/to_ir.lua

```diff
@@ -87,26 +106,44 @@ function ToIR:exit_function(cmds)
     end
 end
 
---
---
---
-function ToIR:is_local(decl)
-    if decl._tag == "ast.Decl.Decl" then
-        local loc_id = self.loc_id_of_decl[decl]
-        local glb_id = self.glb_id_of_decl[decl]
+-- Returns a to_ir.Var object containing ID of `decl`.
+function ToIR:resolve_variable(decl)
+    if decl._tag == "ast.Var.Name" then
+        local glb_id = assert(self.glb_id_of_var[decl])
+        return to_ir.Var.GlobalVar(glb_id)
+    end
+    assert(decl._tag == "ast.Decl.Decl")
+
+    local glb_id = self.glb_id_of_decl[decl]
+    if glb_id then
+        return to_ir.Var.GlobalVar(glb_id)
+    end
+
+    local var
+    local i = 1
+    while i <= #self.func_stack do
+        local loc_id = self.func_stack[i].loc_id_of_decl[decl]
         if loc_id then
-            assert(not glb_id)
-            return true, loc_id
+            var = to_ir.Var.LocalVar(loc_id)
+            break
+        end
+        i = i + 1
+    end
+
+    assert(var)
+    for j = i + 1, #self.func_stack do
+        local func = self.func_stack[j].func
+        local u_id
+        if var._tag == "to_ir.Var.LocalVar" then
+            u_id = ir.add_upvalue(func, decl.name, decl._type, ir.Value.LocalVar(var.id))
```

Comment:
Is it possible that we call `add_upvalue` more than once for the same upvalue, if it is used multiple times inside the same function? I couldn't find an if statement guarding against that, but maybe I missed it.

Reply:
You're correct, I had missed it after the refactor. I added a check for that back in, but I'm not sure if this makes the `resolve_variable` method a bit complicated to read. Should we refactor it out to smaller subroutines?

Reply:
Instead of that `is_captured` for loop, can we use `self.upval_id_of_decl` to find out if an upvalue has already been registered?